model_hub/examples/huggingface/translation/translation_trial.py

```diff
@@ -0,0 +1,348 @@
+            source_lang,
+            target_lang,
+            prefix,
+            data_config,
+            padding,
+            max_target_length,
+            examples,
+        ):
+            inputs = [ex[source_lang] for ex in examples["translation"]]
+            targets = [ex[target_lang] for ex in examples["translation"]]
+            inputs = [prefix + inp for inp in inputs]
+            model_inputs = tokenizer(
+                inputs, max_length=data_config.max_source_length, padding=padding, truncation=True
+            )
+
+            # Setup the tokenizer for targets
+            with tokenizer.as_target_tokenizer():
+                labels = tokenizer(
+                    targets, max_length=max_target_length, padding=padding, truncation=True
+                )
+
+            # If we are padding here, replace all tokenizer.pad_token_id in the labels by -100
+            # when we want to ignore padding in the loss.
+            if padding == "max_length" and data_config.ignore_pad_token_for_loss:
+                labels["input_ids"] = [
+                    [(lb if lb != tokenizer.pad_token_id else -100) for lb in label]
+                    for label in labels["input_ids"]
+                ]
+
+            model_inputs["labels"] = labels["input_ids"]
+            return model_inputs
+
+        tokenized_datasets = {}
+
+        for split in ["train", "validation"]:
+            max_target_length = (
+                self.data_config.max_target_length
+                if split == "train"
+                else self.data_config.val_max_target_length
+            )
+            dataset = self.raw_datasets[split]
+            tokenized_datasets[split] = dataset.map(
+                functools.partial(
+                    preprocess_function,
+                    self.tokenizer,
+                    source_lang,
+                    target_lang,
+                    prefix,
+                    self.data_config,
+                    padding,
+                    max_target_length,
+                ),
+                batched=True,
+                num_proc=self.data_config.preprocessing_num_workers,
+                remove_columns=column_names,
+                load_from_cache_file=not self.data_config.overwrite_cache,
+            )
+            hf.remove_unused_columns(self.model, tokenized_datasets[split])
+
+        # Data collator
+        label_pad_token_id = (
+            -100 if self.data_config.ignore_pad_token_for_loss else self.tokenizer.pad_token_id
+        )
+        if self.data_config.pad_to_max_length:
+            collator = transformers.default_data_collator
+        else:
+            collator = transformers.DataCollatorForSeq2Seq(
+                self.tokenizer,
+                model=self.model,
+                label_pad_token_id=label_pad_token_id,
+                pad_to_multiple_of=8 if self.hparams.use_apex_amp else None,
+            )
+        self.collator = lambda x: collator(x).data
```

Comment:
Not exactly sure how this would be done in a nice way outside of nesting this function in the __init__() which isn't a great solution, but since `self.collator` is used in other functions would be nice to have that defined in the __init__(). I recognize that in this code `preprocess_function` is guaranteed to be called before the data loaders so functionally it's fine, but just wanted to point it out.

Reply:
i moved setting the collator to trial initialization.  the collate functions just need to be created after building the tokenizer.